### PR TITLE
Termination cause

### PR DESCRIPTION
**Pull request by vkatsuba**

### CODE REVIEW

**Review by ebengt — APPROVED** [review]
Looks good.
It would be nice with test of more than 'nornal'.

File: rebar.config [hunk 1]
```diff
@@ -12,7 +12,7 @@
 	{netdata, {git, "https://github.com/RoadRunnr/erl_netdata.git", {ref, "cbd6eaf"}}},
 	{gtplib, {git, "https://github.com/travelping/gtplib.git", {branch, "master"}}},
 	{pfcplib, {git, "https://github.com/travelping/pfcplib.git", {branch, "master"}}},
-	{ergw_aaa, {git, "git://github.com/travelping/ergw_aaa", {tag, "3.5.0"}}},
+	{ergw_aaa, {git, "git://github.com/travelping/ergw_aaa", {branch, "master"}}},
```
Comment (vkatsuba) [inline rebar.config:15]: Changed only for testing.

File: src/ggsn_gn.erl [hunk 2]
```diff
@@ -119,7 +119,7 @@ handle_event({call, From}, terminate_context, _State, Data) ->
 
 handle_event({call, From}, {path_restart, Path}, _State,
 	     #{context := #context{path = Path}} = Data) ->
-    close_pdp_context(normal, Data),
+    close_pdp_context(path_restart, Data),
```
Comment (RoadRunnr) [inline src/ggsn_gn.erl:122]: `path_restart` is a bad name for an externally visible the termination cause. This really is a `peer restart`

File: src/ggsn_gn.erl [hunk 3]
```diff
@@ -251,10 +251,10 @@ handle_event(info, _Info, _State, Data) ->
     keep_state_and_data;
 
 handle_event({timeout, context_idle}, stop_session, _state, Data) ->
-    delete_context(undefined, normal, Data);
+    delete_context(undefined, inactivity_timeout, Data);
```
Comment (RoadRunnr) [inline src/ggsn_gn.erl:254]: we might distinguish between U and C idle timeout.

File: src/ggsn_gn.erl [hunk 4]
```diff
@@ -279,17 +279,18 @@ handle_pdu(ReqKey, #gtp{ie = PDU} = Msg, _State,
     ergw_gsn_lib:ip_pdu(PDU, Context, PCtx),
     {keep_state, Data}.
 
+%% User Plane Error Indication
 handle_sx_report(#pfcp{type = session_report_request,
 		       ie = #{report_type := #report_type{erir = 1}}},
 	    _State, Data) ->
-    close_pdp_context(normal, Data),
+    close_pdp_context(session_error, Data),
```
Comment (RoadRunnr) [inline src/ggsn_gn.erl:286]: wrong naming, A GTP Error Indication, mean the remote peer didn't know about the GTP-U tunnel anymore, could be a remote failure or a peer restart.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]
code looks ok, but the names of the termination causes become externally visible, they needs a high level review and agreement.
And some are plainly wrong (e.g. path_restart)

File: src/ergw_prometheus.erl [hunk 5]
```diff
@@ -105,6 +107,11 @@ declare() ->
 				  {labels, [name, type]},
 				  {buckets, [100, 300, 500, 750, 1000]},
 				  {help, "PFCP Request execution time."}]),
+
+    %% Termination cause metrics
+    prometheus_counter:declare([{name, termination_cause_total},
+                  {labels, [name, type]},
+                  {help, "Total number of termination causes"}]),
```
Comment (RoadRunnr) [inline src/ergw_prometheus.erl:114]: @mgumz @fholzhauser shouldn't that include version and remote as tags as well? Maybe even direction and message type (for incoming requests)?

File: src/ggsn_gn.erl (hunk 3 above)
Comment (fholzhauser) [reply src/ggsn_gn.erl:254]: Yes, that's definitely a good idea. We also still need to improve handling different error scenarios from AAA. I'd propose to address those in a next PR.

File: src/ergw_prometheus.erl (hunk 5 above)
Comment (fholzhauser) [reply src/ergw_prometheus.erl:114]: Yes, we could consider adding those, however I'm not sure how it'd work consistently with the variety of sources these terminations can be initiated from. Proably it requires further discussion to work it out.

**Review by RoadRunnr — APPROVED** [review]

**Pull request merged**